### PR TITLE
Add rules with splits

### LINKED ISSUE #1168: [Feature] Split transactions in schedules
### Verified feature request does not already exist?

- [X] I have searched and found no existing issue

### 💻

- [ ] Would you like to implement this feature?

### Pitch: what problem are you trying to solve?

Can we allow schedules to post split transactions? This will make it easier for users to track what's included in a transaction. An example of where this could be useful is if a user has multiple installments that are charged to an account on the same statement date, users will be able to identify which portion of the transaction is for which purchase.

### Describe your ideal solution to this problem

Allow schedules to post split transactions

### Teaching and learning

_No response_

### PR DESCRIPTION
**Pull request by jfdoming**
Closes #1168 

This PR adds the ability to automatically split transactions from rules. Features:
- Add splits to a rule with per-split actions like category, note, etc.
- Run rules-with-splits when normal rules would apply.
- Working mobile view

Not yet implemented (WIP):
- Display splits on rules page
- Display split transactions post-rule-application in rule editor

Previews:
image

image

image

### CODE REVIEW

**Comment by youngcw** [comment]
I like where this is going. I would like this on a few of my schedules that I currently split by hand each month.

It may be for another PR, but being able to set the percent would be nice

**Comment by joel-jeremy** [comment]
> being able to set the percent would be nice

Agreed. Maybe we could expand the set amount rule action to support percent values? But yeah ideally it should be done in a separate PR.

**Comment by jfdoming** [comment]
This PR is now ready for review! In the interest of keeping things intelligible and focusing on the basic implementation first, I'd like to address these non-blocking features already mentioned in a future PR:
- Display splits on the rules page (currently, it just displays a list of all actions irrespective of the split they belong to)
- Display split transactions post-rule-application in the rule editor (might not do this at all. Currently, if you write any other rule with a condition and the rule updates a transaction so it no longer matches the condition, the transaction is no longer displayed, so doing the same for split-rules would be consistent)
- Support custom percentages for splits (possible implementation I'd like your thoughts on: rather than letting users specify custom percentages, which could require extra validation that the percents all total to 100%, we could use relative weights)

File: packages/desktop-client/src/components/modals/EditRule.jsx
Comment (jfdoming) [reply packages/desktop-client/src/components/modals/EditRule.jsx]: Oh, nice catch! that would be the case, but actually this is a bigger problem left over from when I did a small refactor last week. It should be `actions: getActions(actionSplits).map(unparse)` now since `actionSplits` is an array of objects. Updated!

Comment (jfdoming) [reply packages/desktop-client/src/components/modals/EditRule.jsx]: (side note: `getActions` can use `flatMap`, though)

**Comment by joel-jeremy** [comment]
Removing an action (by clicking on `-`) results in a fatal error

**Comment by jfdoming** [comment]
> Removing a rule action (by clicking on `-`) results in a fatal error

haha I just discovered this as well... fixed!

**Comment by joel-jeremy** [comment]
Posting a scheduled split transaction doesn't seem to work for me.

Steps to reproduce:
1. Create a schedule with splits
2. Go to transactions page
3. Manually post the scheduled transaction

**Comment by jfdoming** [comment]
> Posting a scheduled split transaction doesn't seem to work for me.
> 
> Steps to reproduce:
> 
> 1. Create a schedule with splits
> 2. Go to transactions page
> 3. Manually post the scheduled transaction

Huh, interesting. I'll add this to the queue of bugs I'm fixing (currently trying to fix the handling of empty splits)

**Comment by jfdoming** [comment]
@joel-jeremy the bug you found (and several others) should be fixed with the latest commits I've added. Would you mind taking a look when you have a moment? I tried to test as many features as I am aware of, but it can be hard to check everything.

To be honest, given the number of potential repercussions to enabling this feature, I'd be open to marking this as experimental and adding an option gating the Rules-page "Split" button temporarily. Thoughts? That way, I could also make a separate PR to add a proper UI to the rules page, which is usable but currently looks a bit broken:
image
image

I'm fine with just adding that UI in this PR as well, depends on how much code you want to review at once!

**Comment by youngcw** [comment]
It may be good to set this as experimental since there are probably a number of edge cases that we haven't run into yet.

**Comment by jfdoming** [comment]
> looks like percentages are not getting handled properly. Its happening both on a early schedule post and by entering in the needed values manually. The other data fields seem to work right

Oh that's odd, I'll look into it. Agree with experimental, I'll add an option.

**Comment by joel-jeremy** [comment]
A few more observations:

- I think we should limit some of the `set` choices in the split rule:

image

Users should only be able to set `category`, `payee`, and `notes` for split entries.

- When adding a transaction with an applied split rule and try editing the parent transaction, an error occurs:
image

**Comment by youngcw** [comment]
Should there be some sanity check on the split amounts?  Its currently possible to over or under allocate the split.  Thats fine for manual entries, but bad for imports.

Some checks like:
* Is there an allocate remainder line
* Do percentages add up to 100 if all lines are percents
* Do amounts add up for exact amount matching

Or maybe conditionally add an extra split line to handle balancing out.

**Comment by jfdoming** [comment]
> Should there be some sanity check on the split amounts? Its currently possible to over or under allocate the split. Thats fine for manual entries, but bad for imports.
> 
> Some checks like:
> 
> * Is there an allocate remainder line
> * Do percentages add up to 100 if all lines are percents
> * Do amounts add up for exact amount matching
> 
> Or maybe conditionally add an extra split line to handle balancing out.

I believe I configured these rules to not match if there are only fixed amounts and they don't have the correct total. Do you think that behaviour is too confusing?

**Comment by joel-jeremy** [comment]
> Or maybe conditionally add an extra split line to handle balancing out.

I think this one makes the most sense to me and minimizes the complexity. If there is any remainder, the rule should just add another empty split which be populated by the user.

One common scenario where I see myself using this split rule is automatically adding transaction fees and having the rule auto-populate the fees, I only need to populate the empty transaction that the rule created with the actual transaction amount.

**Comment by youngcw** [comment]
I agree that adding in the extra line if there is leftover makes the most sense and is probably the easiest for users to work around.

**Comment by jfdoming** [comment]
@youngcw @joel-jeremy thanks for your patience as I worked through the list of bugs and blocking features! I believe this PR should be ready for review again. Did some more bashing on the transactions page to fix a few extra straggler bugs so I'd love to get your fresh eyes on it!

Given that there's now a feature flag, I think it makes sense to defer mobile support and the rules display page to future PRs.

**Comment by youngcw** [comment]
Nice.  This seems to be working as I would expect now.

My only gripe now is that I keep putting in positive values expecting them to be payments in transaction.  Maybe you could use the amount picker that the schedules use to help that?  Its not a big deal since that matches current rule function

**Comment by jfdoming** [comment]
> Nice. This seems to be working as I would expect now.
> 
> My only gripe now is that I keep putting in positive values expecting them to be payments in transaction. Maybe you could use the amount picker that the schedules use to help that? Its not a big deal since that matches current rule function

That's a great point and something I've run into a few times actually! I'll see what I can come up with.

**Comment by jfdoming** [comment]
> Nice. This seems to be working as I would expect now.
> 
> My only gripe now is that I keep putting in positive values expecting them to be payments in transaction. Maybe you could use the amount picker that the schedules use to help that? Its not a big deal since that matches current rule function

Just to confirm, were you thinking to change _all_ rule amount inputs to use the schedule picker?

**Comment by youngcw** [comment]
> Just to confirm, were you thinking to change _all_ rule amount inputs to use the schedule picker?

That would be nice, but that might be for a different PR where all the amount lines change

**Comment by jfdoming** [comment]
> > Just to confirm, were you thinking to change _all_ rule amount inputs to use the schedule picker?
> 
> That would be nice, but that might be for a different PR where all the amount lines change

Yeah, I'm in favour of a different PR given the size of this one. (I think it would be easiest to do all at once.) To summarize (mainly so I don't forget myself), here are the development pieces that are still needed as far as I know:
- Confirm/fix behaviour on mobile
- Add support to the rules display page
- Use a better amount input on the rule entry page

**Review by joel-jeremy — APPROVED** [review]
LGTM! One gripe that I have is that when the split rule gets executed before the user has filled up the amount field, the amounts don't recalculate when the user does update the amount. That is a possible future enhancement that we can introduce though and is out of scope of this PR. Thank you!

**Comment by jfdoming** [comment]
> LGTM! One gripe that I have is that when the split rule gets executed before the user has filled up the amount field, the amounts don't recalculate when the user does update the amount. That is a possible future enhancement that we can introduce though and is out of scope of this PR. Thank you!

Awesome, thanks for the approval! Is there someone else I should get to take a look before we merge? (I don't have write permissions so I can't merge it myself)

**Comment by glp-66** [comment]
dont know if I should add this here or not. I have a schedule recurring on the 15th for $150. It has 2 splits built in the rules. $100 and $50. different categories. I posted the split in my account where the split is, and  it comes up as missed still. I still then need to go into schedules and skip it. even trying to post it in the schedule. doesn't do anything in the schedule, but it posts the transaction.
Thanks for working on this feature though. 1 thing I was really looking to have implemented.

**Comment by glp-66** [comment]
> > dont know if I should add this here or not. I have a schedule recurring on the 15th for $150. It has 2 splits built in the rules. $100 and $50. different categories. I posted the split in my account where the split is, and it comes up as missed still. I still then need to go into schedules and skip it. even trying to post it in the schedule. doesn't do anything in the schedule, but it posts the transaction. Thanks for working on this feature though. 1 thing I was really looking to have implemented.
> 
> Would you be able to share a budget file? Feel free to remove transactions/change numbers for privacy, just looking to repro your issue! (Also FYI there's #2379 for bugs/requests with this feature)

I just did a quick test on a backup file and changed the dates and it seemed to work. My good file has it to happen on the 15th so I'll see how that triggers and will update #2379 if it does not mark the schedule as paid. Thanks!

**Pull request merged**
